rspace/src/test/scala/coop/rchain/rspace/test/InMemoryStore.scala

```diff
@@ -2,185 +2,186 @@ package coop.rchain.rspace.test
+  private[rspace] override def hashChannels(channels: Seq[C]): H =
+    Codec[Seq[C]]
+      .encode(channels)
+      .map((bitVec: BitVector) => Blake2b256Hash.create(bitVec.toByteArray))
+      .get
 
-  private[rspace] def createTxnRead(): Unit = ()
+  private[rspace] override def createTxnRead(): Unit = ()
 
-  private[rspace] def createTxnWrite(): Unit = ()
+  private[rspace] override def createTxnWrite(): Unit = ()
 
-  private[rspace] def withTxn[R](txn: T)(f: T => R): R =
+  private[rspace] override def withTxn[R](txn: T)(f: T => R): R =
     f(txn)
 
-  private[rspace] def collectGarbage(txn: T,
-                                     channelsHash: H,
-                                     dataCollected: Boolean = false,
-                                     waitingContinuationsCollected: Boolean = false,
-                                     joinsCollected: Boolean = false): Unit =
-    collectGarbage(channelsHash)
-
-  private[this] def collectGarbage(key: H): Unit = {
-    val as = _data.get(key).exists(_.nonEmpty)
-    if (!as) {
-      //we still may have empty list, remove it as well
-      _data.remove(key)
-    }
+  private[rspace] override def getChannels(txn: T, s: H) =
+    _dbGNATs.get(s).map(_.channels).getOrElse(Seq.empty)
 
-    val psks = _waitingContinuations.get(key).exists(_.nonEmpty)
-    if (!psks) {
-      //we still may have empty list, remove it as well
-      _waitingContinuations.remove(key)
-    }
+  private[rspace] override def getData(txn: T, channels: Seq[C]): Seq[Datum[A]] =
+    _dbGNATs.get(hashChannels(channels)).map(_.data).getOrElse(Seq.empty[Datum[A]])
 
-    val cs    = _keys.getOrElse(key, Seq.empty[C])
-    val joins = cs.size == 1 && _joinMap.contains(cs.head)
+  private[this] def getWK(txn: T, curr: Seq[C]): Seq[WaitingContinuation[P, K]] =
+    _dbGNATs
+      .get(hashChannels(curr))
+      .map(_.wks)
+      .getOrElse(Seq.empty[WaitingContinuation[P, K]])
 
-    if (!as && !psks && !joins) {
-      _keys.remove(key)
-    }
-  }
+  private[rspace] override def getWaitingContinuation(
+      txn: T,
+      curr: Seq[C]): Seq[WaitingContinuation[P, K]] =
+    getWK(txn, curr)
+      .map { (wk: WaitingContinuation[P, K]) =>
+        wk.copy(continuation = InMemoryStore.roundTrip(wk.continuation))
+      }
 
-  private[rspace] def putDatum(txn: T, channels: Seq[C], datum: Datum[A]): Unit = {
-    val key = hashChannels(channels)
-    putCs(txn, channels)
-    val datums = _data.getOrElseUpdate(key, Seq.empty[Datum[A]])
-    _data.update(key, datum +: datums)
+  override def getPatterns(txn: T, channels: Seq[C]): Seq[Seq[P]] =
+    getWK(txn, channels).map(_.patterns)
+
+  private[rspace] override def getJoin(txn: T, c: C): Seq[Seq[C]] =
+    _dbJoins.getOrElse(c, Seq.empty[Seq[C]])
+
+  private[this] def withGNAT(txn: T, key: H)(
+      f: (T, H, Option[GNAT[C, P, A, K]]) => Option[GNAT[C, P, A, K]]): Unit = {
+    val gnatOpt = _dbGNATs.get(key)
+    val rOpt    = f(txn, key, gnatOpt)
```

Comment:
😎 😎 😎 

Seems like `txn` and `key` are usually in scope in a context where yr calling `withGNAT` (so long as yr not inlining the call to `hashChannels`).
 
What's the motivation for explicitly threading them through?

Could we simplify and have the signature be something like:
```scala
private[this] def withGNAT(txn: T, key: H)(f: Option[GNAT[C, P, A, K]] => Option[GNAT[C, P, A, K]]): Unit
```

Reply:
agreed that's odd.
I'm estimating that a `withGNAT` will have be opening new txn and passing them onwards.

I'll kill it here and introduce it when (if) it's needed